src/main/scala/com/fulcrumgenomics/umi/UmiConsensusCaller.scala

```diff
@@ -0,0 +1,277 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.umi
+
+import com.fulcrumgenomics.FgBioDef._
+import com.fulcrumgenomics.umi.UmiConsensusCaller._
+import htsjdk.samtools.SAMFileHeader.{GroupOrder, SortOrder}
+import htsjdk.samtools.util.{Murmur3, SequenceUtil}
+import htsjdk.samtools.{SAMFileHeader, SAMReadGroupRecord, SAMRecord, SAMTag}
+
+import scala.collection.JavaConversions.iterableAsScalaIterable
+import scala.collection.mutable
+
+/**
+  * Contains shared types and functions used when writing UMI-driven consensus
+  * callers that take in SAMRecords and emit SAMRecords.
+  */
+object UmiConsensusCaller {
+  /** The type of consensus read to output. */
+  object ReadType extends Enumeration {
+    type ReadType = Value
+    val Fragment, FirstOfPair, SecondOfPair = Value
+  }
+
+  /**
+    * Stores information about a read to be fed into a consensus.
+    */
+  case class SourceRead(id: String, bases: Array[Byte], quals: Array[Byte]) {
+    require(bases.length == quals.length, "Bases and qualities are not the same length.")
+    def length: Int = bases.length
+  }
+
+  /** A trait that consensus reads must implement. */
+  trait ConsensusRead {
+    /** The ID of the molecule that generated the consensus read. */
+    def id: String
+    /** The bases of the consensus read. */
+    def bases: Array[Byte]
+    /** The quals of the consensus read. */
+    def quals: Array[Byte]
+    /** Gets the length of the consensus read. */
+    def length: Int = bases.length
+    /** Returns the consensus read a String - mostly useful for testing. */
+    def baseString = new String(bases)
```

Comment:
Done.  I've renamed the trait to `SimpleRead` because I felt like `Read` was too generic, and it's is incredibly simple 😛